### ISSUE TITLE
Chado CV/DB table challenge

### ISSUE THREAD

**Issue by spficklin** [GMOD/Chado#58]
We are having a bit of trouble organizing controlled vocabularies in the CV and DB tables.  The problem comes in with the "short name" (e.g. GO, SO, etc...) for the vocabulary which gets stored in the db.name column.  The db table also houses the URL and URL prefix.  The URL prefix is meant to be used for any dbxref (even if it's not a term) to rebuild a web link to point to the cross reference.  

This works great for the Gene Ontology because all of the terms have one short name (GO), so that record get stores in the db table and the three vocabularies for GO (i.e. molecular_function, biological_process, cellular_component) all get stored in in the cv table.   This works great!   The db.urlprefix can be used to build all of the link-outs because GO uses the same URL for all of them.

This breaks for vocabularies such as EDAM.  "EDAM is an ontology of bioinformatics types of data including identifiers, data formats, operations and topics." In this case the entire vocabulary is named 'EDAM' and it has separate vocabularies (like GO) named 'data', 'operation',  'format', etc.  The terms in this ontology are not prefixed with 'EDAM' but rather the vocabulary name.   So, to make link-outs work, this forced us to reverse how we store them.  We put 'EDAM' in the cv table... which is not ideal.

KEGG provides more challenges.   We want to store KEGG orthologs and pathway as cvterms.  These terms are organized into KEGG's BRITE hierarchy.  So, in a way it seems like an ontology.  Because orthologs and pathways have different URLs we are forced to store the PATH and KO vocabularies in the db table.   But, if we want to preserve the organization of the BRITE heirarchy we have to lump these into the same KEGG BRITE cv record.  Same problem as EDAM, except that higher level terms in the BRITE hierarchy don't have corresponding pages so we can't link to them.  Some of our term links will be broken.

So, I think we need to adjust how we recommend and how our scripts store vocabularies in Chado.   Here's one suggestion

1) We keep the db.urlprefix field but use that as a "default".  This ensures continued backwards compatibility.
2) Because a database can use multiple URLs for all kinds of data (not just terms) and even for different vocabularies, we create a new dburl table where URLs can be stored.  We document a storage format that all Chado client applications can understand if they follow it.  For example, Tripal currently uses {db} and {accession} as tokens in the URL string. This allows the URL to be constructed by filling in the tokens and those tokens can appear anywhere in the string. Whereas now, usage implies that the accession is added to the end of the value in the db.urlprefix.
3)  We create a new dbxref_dburl table that allows a dbxref record to specify a different URL to use.  So, if one is not specified then the default urlprefix is used.  

This approach allows us to have only one true database record, and put the CV's where they go in the cv table.  It would allow us to have multiple URLs per database, and indicate a different URL construction string if needed.   This would solve our KEGG problem as well because we could leave the db.urlprefix blank and add to the dburl table specific URLs for dbxref records that use them.

Any thoughts?

**Comment by spficklin** [comment]
Oh, I just realized that the short_name for the accession can be different too, so rather than a dburl table what if we have a dbnamespace table.  We can then include the desired short name for the vocabulary and the url constructor string,  and a dbxref_namespace linker table

**Comment by ekcannon** [comment]
I like the idea of **dbnamespace** and **dbxref_namespace** tables, even in the absence of multiple namespaces within an ontology as it makes it very clear where the namespace prefix comes from. It took me a little while to figure out that db.name held the namespace prefix when first learning chado. 

Are there still problems with multiple URLs within a single ontology? If so, would the URL be a field in the **dbnamespace** table? Optional? Required? This could add some confusion for new developers if urlprefixes can be set in two places.

**Comment by spficklin** [comment]
Yesh. EDAM is a good example. It has multiple namespaces for it's different sub vocabularies.

I do agree it can get confusing to have the dd.urlprefix and the dbnamespace table. I was hoping to suggest something that would bbe backwards compatible. Perhaps if we mark the **db.urlprefix** as deorecated and encourage use of the new **dbnamespace** table. 

Would you be okay with that approach @ekcannon ?

I would love to fix this problem for Tripal v3. @scottcain any thoughts?

**Comment by ekcannon** [comment]
I'm thinking more of deprecating the use of db.name for a namespace (but not the field itself). db.urlprefix is also used to build URLs for things like DOIs, GenBank, and GRIN accessions, so db.urlprefix is still essential.

My question is whether a single ontology might have multiple URL prefixes, or if all terms can be linked with one URL prefix+namespace+accession, using db.urlprefix. In other words, would urlprefix be needed in the table **dbnamespace**.
<pre>
  CREATE TABLE dbnamespace (
    dbnamespace_id BIGSERIAL NOT NULL, 
    namespace VARCHAR(32) NOT NULL,
    urlprefix VARCHAR(255)
  );</pre>

**Comment by spficklin** [comment]
The word **namespace** I think is a bit confusing.  On the gene ontology website they refer to their three vocabularies as different namespaces despite they all use the same **short name:** GO.  In the semantic, of which vocabularies form a major component,  a namespace is the the url prefix that uniquely identifies terms in a vocabulary. 

So in the case of EDAM it does use different URL namespaces for a variety of vocabularies even though it's short name is EDAM. 

So what about this proposal

| Table | Column | Comment |
| -------- | ------------- | ---------------- |
| db | db_id | Primary key |
| db | name | The name of the database.  In previous versions of  Chado this field stored the database abbreviation used to build the namespace URL for an accession in the database. Now this field should store the full name of the database |
| db | short_name | The abbreviated name for the database.  For example: GO, SO, etc. |
| db | url | The full URL for home page of this database on the web. |
| db | urlprefix | The URL template used to construct the namespace for this database.  You can use tokens to identify where specific values should go.  These include: {db} for the database short name and {accession} for a resources' unique identifier. If a database maintains multiple namespaces you can add more to the dbnamespace table and link dbxref entries to the appropriate one using the dbxref_namespace table  |

| Table | Column | Comment |
| -------- | ------------- | ---------------- |
| dbnamespace | dbnamespace_id | Primary key |
| dbnamespace | db_id | Foreign key to a db record |
| dbnamespace | short_name | The sshort name used to construct the unique namespace URL |
| dbnamespace | urlprefix | These same as for the db.urlprefix column |

**Comment by spficklin** [comment]
An alternate solution to this problem may be a ``db_relationship`` table.  It would allow each **ontology ID prefix** (what I was calling a **short name** above) to continue to be stored in the ``db`` table.  But it would the allow us to store the parent database record.  So we would continue to store EDAM's prefixes (``data``, ``operation``, ``format``, etc) in the ``db`` table, but we could then add a new ``EDAM`` record to the ``db`` table and link the ``data``, ``operation``, ``format`` as a child of the ``EDAM`` record.   

This solution preserves backwards compatibility with Chado v1.3 and adds one table that follows the same structure as other relationship tables (so not hard to deal with).   Client tools could then recognize when multiple ontologies belong together and use the parent record name when referring to them all.

**Comment by spficklin** [comment]
As another example, this type of relationship table would allow us to maintain different records for NCBI databases (e.g. SRA, Taxonomy, etc.) and link them all together under an NCBI record.  

Here's some SQL for creating such a table:

```
-- ================================================
-- TABLE: db_relationship
-- ================================================
create table db_relationship (
    db_relationship_id bigserial not null,
    type_id bigint not null,
    subject_id bigint not null,
    object_id bigint not null,
    primary key (db_relationship_id),
    foreign key (type_id) references db (db_id) on delete cascade INITIALLY DEFERRED,
    foreign key (subject_id) references db (db_id) on delete cascade INITIALLY DEFERRED,
    foreign key (object_id) references db (db_id) on delete cascade INITIALLY DEFERRED,
    constraint db_relationship_c1 unique (subject_id,object_id,type_id)
);
create index db_relationship_idx1 on db_relationship USING btree (type_id);
create index db_relationship_idx2 on db_relationship USING btree (subject_id);
create index db_relationship_idx3 on db_relationship USING btree (object_id);

COMMENT ON TABLE db_relationship IS 'Specifies relationships between databases.  This is 
particularly useful for ontologies that use multiple prefix IDs for it's vocabularies. For example, 
the EDAM ontology uses the prefixes "data", "format", "operation" and others. Each of these would 
have a record in the db table.  An "EDAM" record could be added for the entire ontology to the 
db table and the previous records could be linked as "part_of" EDAM.  As another example
databases housing cross-references may have sub databases such as NCBI (e.g. Taxonomy, SRA, etc).
This table can use a 'part_of' record to link all of them to NCBI.

```

**Comment by scottcain** [comment]
Is this idea mature enough for 1.4? It feels like it.

**Comment by bradfordcondon** [comment]
I am marking this issue as **high** priority as we have several tripal issues contingent on this.

I'm marking the difficulty low because @spficklin has kindly provided a draft schema.

**Comment by spficklin** [comment]
Hi @scottcain when you ask if this feels mature enough for 1.4 do you mean the latter solution of a `db_relationship` table?  I'm not 100% sure we've captured all the use cases, but it certainly resolves a lot of problems.

**Comment by spficklin** [comment]
And it falls within the typical  Chado design style so very low cost to include IMHO.

**Comment by ekcannon** [comment]
Echoing @spficklin's comment: could the full proposal be listed here? It isn't clear to me what tables and table designs made it through the discussion filter.

**Comment by spficklin** [comment]
@ekcannon I proposed both solutions but for me I think I'm leaning towards just a adding a `db_relationship` table and using relationships between them to make sense of vocabularies that have different prefixes but fall within a single group.    Storing ontologies in Chado has become very confusing lately for multiple reasons.  So, what may be best is before we implement changes to any of the cv, cvterm, db, dbxref tables that we implement a loader with any proposed changes and load a set of "problematic" ontologies to make sure anything we come up with works.  @bradfordcondon and I were discussing this related to KEGG ontologies so I can volunteer my group to do it.  Although, I don't have a timeline for when we could get it done.

**Comment by cmungall** [comment]
Not sure if this helps, but the OBO community is moving towards publishing `-base` versions of some ontologies which are guarantee to include only classes that belong to that ontology (and possibly edges that point to classes in other ontologies).

I seem to recall answering a question on this somewhere else, maybe in the context of pombase, @kimrutherford does this ring a bell?
